src/main/java/frc/robot/subsystems/shooter/Shooter.java

```diff
@@ -0,0 +1,232 @@
+
+  private int topAtSetpointIterationCount = 0;
+  private int bottomAtSetpointIterationCount = 0;
+  private int angleAtSetpointIterationCount = 0;
+
+  public Shooter(ShooterIO io, Drivetrain drivetrain, Intake intake) {
+    this.io = io;
+    this.intake = intake;
+    this.angleTreeMap = new InterpolatingDoubleTreeMap();
+    this.veloTreeMap = new InterpolatingDoubleTreeMap();
+
+    this.autoShooter = false;
+    this.hasNote = false;
+    this.isAimed = false;
+    this.aiming = false;
+    this.scoringAmp = false;
+    this.scoringSubwoofer = false;
+    this.scoringPodium = false;
+    this.inStorage = () -> false;
+
+    if (TESTING) {
+      ShuffleboardTab tab = Shuffleboard.getTab(SUBSYSTEM_NAME);
+      tab.add(SUBSYSTEM_NAME, this);
+    }
+  }
+
+  @Override
+  public void periodic() {
+    io.updateInputs(shooterInputs);
+    Logger.processInputs(SUBSYSTEM_NAME, shooterInputs);
+    this.inStorage =
+        () -> (Math.abs(shooterInputs.angleEncoderAngleDegrees) < ShooterConstants.ANGLE_TOLERANCE);
+    // this.hasNote = intake.hasNote();
+
+    if (TESTING) {
+      io.setShooterWheelBottomVelocity(bottomWheelVelocity.get());
+      io.setShooterWheelTopVelocity(topWheelVelocity.get());
+      io.setAngle(angle.get());
+    } else if (autoShooter == false) {
+      io.setShooterWheelBottomVelocity(SHOOTER_IDLE_VELOCITY);
+      io.setShooterWheelTopVelocity(SHOOTER_IDLE_VELOCITY);
+    } else {
+      this.runAngleStateMachine();
+      io.setShooterWheelBottomVelocity(SHOOTER_IDLE_VELOCITY);
+      io.setShooterWheelTopVelocity(SHOOTER_IDLE_VELOCITY);
+    }
+  }
+
+  public void updateAlliance(Alliance newAlliance) {
+    this.alliance = newAlliance;
+  }
+
+  private void runAngleStateMachine() {
+    double test =
+        Field2d.getInstance()
+            .getAllianceSpeakerCenter()
+            .minus(RobotOdometry.getInstance().getEstimatedPosition())
+            .getTranslation()
+            .getNorm();
+    if (hasNote == true) {
+      if (this.inRange() == true) {
+        if (scoringPodium == true) {
+          io.setShooterWheelBottomVelocity(ShooterConstants.PODIUM_VELOCITY);
+          io.setShooterWheelTopVelocity(ShooterConstants.PODIUM_VELOCITY);
+          io.setAngle(ShooterConstants.PODIUM_ANGLE);
+        } else if (scoringSubwoofer == true) {
+          io.setShooterWheelBottomVelocity(ShooterConstants.SUBWOOFER_VELOCITY);
+          io.setShooterWheelTopVelocity(ShooterConstants.SUBWOOFER_VELOCITY);
+          io.setAngle(ShooterConstants.SUBWOOFER_ANGLE);
+        } else if (scoringAmp == true) {
+          io.setShooterWheelBottomVelocity(ShooterConstants.AMP_VELOCITY);
+          io.setShooterWheelTopVelocity(ShooterConstants.AMP_VELOCITY);
+          io.setAngle(ShooterConstants.AMP_ANGLE);
+        } else if (aiming == true) {
+          io.setShooterWheelBottomVelocity(
+              veloTreeMap.get(RobotOdometry.getInstance().getEstimatedPosition().getX()));
```

Comment:
the interpolation table is a mapping from distance to the speaker to velocity (or angle); therefore, you have to pass the distance to the speaker (based on the current pose) to the method; not the position on the field